### ISSUE TITLE
can not create ids with no normalizations for certain cgyro runs. 

### ISSUE THREAD

**Issue by jmcclena** [bpatel2107/pyrokinetics#334]
branch: `enhancements/imas_update_v3.4`

I am trying to generate an ids with no dimensional information. If I take the `input.cgyro` from https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/src/pyrokinetics/templates/input.cgyro, reduce max_time and run cgyro on it. Then run the following script

```
import json
import numpy as np
from pyrokinetics import Pyro,template_dir
from pyrokinetics.databases.imas import pyro_to_imas_mapping
import idspy_toolkit as idspy
from idspy_dictionaries import ids_gyrokinetics_local as gkids
from pathlib import Path

cgyro_dir  = Path("/Users/mcclenaghan/programming/python/pyrokinetics/tmp")
pyro = Pyro(gk_file=cgyro_dir / "input.cgyro", gk_code="CGYRO")


pyro.load_gk_output()

ids = gkids.GyrokineticsLocal()
idspy.fill_default_values_ids(ids)
ids = pyro_to_imas_mapping(
        pyro,
        comment=f"Testing IMAS CGYRO",
        ids=ids
    )

```

I get the following normalization error:

```

---------------------------------------------------------------------------
PyroNormalisationError                    Traceback (most recent call last)
Cell In[5], line 24
     22 gkdict = gkids.GyrokineticsLocal()
     23 idspy.fill_default_values_ids(gkdict)
---> 24 ids = pyro_to_imas_mapping(
     25         pyro,
     26         comment=f"Testing IMAS CGYRO",
     27         ids=gkdict
     28     )

File [~/miniconda3/lib/python3.11/site-packages/pyrokinetics/databases/imas.py:210](http://localhost:8888/lab/tree/~/miniconda3/lib/python3.11/site-packages/pyrokinetics/databases/imas.py#line=209), in pyro_to_imas_mapping(pyro, comment, time_interval, reference_values, ids)
    207 original_convention = norms.default_convention
    209 if pyro.gk_output:
--> 210     pyro.gk_output.to(norms.imas)
    211     original_theta_output = pyro.gk_output["theta"].data
    212 else:

File [~/miniconda3/lib/python3.11/site-packages/pyrokinetics/gk_code/gk_output.py:816](http://localhost:8888/lab/tree/~/miniconda3/lib/python3.11/site-packages/pyrokinetics/gk_code/gk_output.py#line=815), in GKOutput.to(self, norms)
    804 """
    805 
    806 Parameters
   (...)
    813 GKOutput with units from norms
    814 """
    815 for data_var in self.data_vars:
--> 816     self.data[data_var].data = self.data[data_var].data.to(norms)
    818 # Coordinates with units not supported in xarray need to manually change
    819 new_coords = {}

File [~/miniconda3/lib/python3.11/site-packages/pyrokinetics/units.py:115](http://localhost:8888/lab/tree/~/miniconda3/lib/python3.11/site-packages/pyrokinetics/units.py#line=114), in PyroQuantity.to(self, other, *contexts, **ctx_kwargs)
    113         as_physical = self._convert_simulation_units(other)
    114         value = as_physical.to(self._convert_base_units(other))
--> 115         return self._replace_nan(value, other)
    117 return super().to(other, *contexts, **ctx_kwargs)

File [~/miniconda3/lib/python3.11/site-packages/pyrokinetics/units.py:54](http://localhost:8888/lab/tree/~/miniconda3/lib/python3.11/site-packages/pyrokinetics/units.py#line=53), in PyroQuantity._replace_nan(self, value, system)
     51 # If everything is a NaN then conversion failed otherwise some
     52 # NaNs exist in the data and we can proceed
     53 if np.isnan(value).all():
---> 54     raise PyroNormalisationError(system, self.units)
     55 else:
     56     return value

PyroNormalisationError: Cannot convert 'dimensionless' to 'imas_input0003' normalisation. Possibly 'imas_input0003' is missing physical reference values. You may need to load a kinetics or equilibrium file
```

However, if I take https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/src/pyrokinetics/templates/outputs/CGYRO_linear/input.cgyro, run cgyro with it, and run the script, then it produces the ids just fine.

**Comment by jmcclena** [comment]
Weird.  It seems that the variable that causes this difference in behavior is `DELTA_T`.

**Comment by bpatel2107** [comment]
The issue above happens when converting the GKOutputs, not the GKInputs. `DELTA_T` has units of `lref/vref` so it shouldn't be dimensionless to begin with. 

Even so pyro should have no issue converting dimensionless quantities as they are dimensionless.

In any case I can't reproduce your error with the script above. The confusing thing here is that your `norms.imas` has "physical units" (you can tell as the name of the unit is `imas_input0003` and not just `imas`) which means at some point in your script you define a unit somewhere and then you likely don't define any others. Can you show me the exact script and CGYRO input file you used? 

The error comes up when it tries to convert something and all the converted values are `NaNs`.  This can happen either when the conversion fails or all the original data are NaNs. So there is a small chance that all your data somehow a NaN but it seems unlikely unless its a weird time step issue causing a bunch of NaNs.

You can also just do the following without needing to import anything from `idspy_toolkit`

```python
from pyrokinetics import Pyro
from pyrokinetics.databases.imas import pyro_to_ids

pyro = Pyro(gk_file="input.cgyro", gk_code="CGYRO")
pyro.load_gk_output()

ids = pyro_to_ids(pyro, comment="Testing IMAS CGYRO")
```

Basically it will only write the HDF5 file if you specify `file_name` in `pyro_to_ids()`. More documentation on this will eventually come.

**Comment by jmcclena** [comment]
The exact script (script.py) and cgyro files script I'm using: 

[mcclenaghan_units_error.zip](https://github.com/pyro-kinetics/pyrokinetics/files/15276276/mcclenaghan_units_error.zip)

**Comment by bpatel2107** [comment]
Ah that is sneaky! This is failing because `growth_rate_tolerance` is a NaN. This is because it uses the final 80% of your data to see of the run has converged or not and that run only has 4 time points. This means that the tolerance on the growth rate only uses 1 point. So the following lines

https://github.com/pyro-kinetics/pyrokinetics/blob/b91f7d97cb8d223a7f0d0be3d080689738d25bff/src/pyrokinetics/gk_code/gk_output.py#L654-L657

result in a divide by 0. Not really a fault on your end but the fix would be to check if all the data was a NaN to begin with

**Comment by bpatel2107** [comment]
@jmcclena Pushed a fix that checks if the data was originally a NaN. Should be working now as I can run your script without issue with the fix.

**Comment by jmcclena** [comment]
Thanks! It's working now.
